File: Hard/84/84.py
# @Date    : 19:37 03/16/2021
# @Author  : ClassicalPi
# @FileName: 84.py
# @Software: PyCharm
import sys
class Solution:
    def largestRectangleArea(self, height):
        height.append(0)
        stack = [-1]
        ans = 0
        for i in range(len(height)):
            while height[i] < height[stack[-1]]:
                h = height[stack.pop()]
                w = i - stack[-1] - 1
                ans = max(ans, h * w)
            stack.append(i)
        height.pop()
        return ans
if __name__ == '__main__':
    S=Solution()
    print(S.largestRectangleArea([2,1,5,6,2,3]))
    print(S.largestRectangleArea([2, 4]))
    print(S.largestRectangleArea([2,1,2]))
    print(S.largestRectangleArea([1,1]))

File: GoVersion/Hard/84.go
package main

import (
	"fmt"
	"math"
)

func largestRectangleArea(heights []int) int {
	var stack [] int
	stack=append(stack, 0)
	heights=append(heights,0)
	res:=0

	for i:=0;i<len(heights);i+=1{
		if len(stack)>1{
			j:=stack[len(stack)-1]
			for j>=0 && heights[i]<heights[j]{
				height:=heights[j]
				stack=stack[:len(stack)-1]
				width:=i-stack[len(stack)-1]-1
				res= int(math.Max(float64(res), float64(height*width)))
				fmt.Println(height,width,res)
				j=stack[len(stack)-1]
			}
		}
		stack=append(stack,i)
	}
	return res
}


